Added Blackbox Erase to a switch

brycedjohnson opened the pull request:
I was always forgotting to erase the log before I was out on the field, so added a blackbox erase to a switch.

Only works with chip flash. 

Probably need a way to tell someone that it will take 20-30 secs to erase.

Haven't yet tested in the air but appears to work on the bench.  I wouldn't merge until some real world testing.  

It should work to: start the erase -> hit the switch to start blackbox (but it won't actually start until erase is done -> fly for 20-30s -> newly erased blackbox will start to log.

Input?

blckmn commented:
If we had some kind of basic file system, we could simply nuke the file system index

borisbstyle commented:
I am not sure if this feature should be introduced as it can take quite some while before erasing finished.

Also quad should not be arming during erase

mikeller commented:
@brycedjohnson: I like the idea. One thing that I'd like to see added is a confirmation beeep once the erase is complete.

@borisbstyle: While it can take a while, I still think being able to make room for a blackbox log while in the field is valuable. But totally agreed on not arming while the erase is ongoing.

ledvinap commented on src/main/fc/fc_msp.c:
```diff
@@ -144,10 +144,11 @@ static const box_t boxes[CHECKBOX_ITEM_COUNT + 1] = {
-    { BOX3DDISABLESWITCH, "DISABLE 3D SWITCH;", 29},
-    { BOXFPVANGLEMIX, "FPV ANGLE MIX;", 30},
+    { BOXBLACKBOXERASE, "BLACKBOX ERASE;", 27 },
+    { BOXFAILSAFE, "FAILSAFE;", 28 },
```
permanent IDs are altered here, it will probably break a lot of things

brycedjohnson replied:
Yeah I wasn't sure if that should be grouped with like items or just added on the end.   The end would be better?

ledvinap commented:
@blckmn : The flash must be physically erased (so that erase does not block write), file system won't help

brycedjohnson commented:
@mikeller: A confirmation beep is a good idea.

@borisbstyle: It looks like while the erasing is going on it shouldn't be blocking on anything, so it shouldn't be a problem to be able to arm while the erase is going on.  I put a check here:

https://github.com/betaflight/betaflight/pull/2267/files#diff-2a14bcfc66c33cd4eab81b64ee4fdae6R800

for the erasing state.  What should happen it is should periodically checking if the erase is done yet in the blackbox state machine.  When you do the erase the states should go like this

BLACKBOX_STATE_STOPPED -> BLACKBOX_STATE_START_ERASE -> BLACKBOX_STATE_ERASING -> BLACKBOX_STATE_PREPARE_LOG_FILE (maybe this should go back to BLACKBOX_STATE_STOPPED)

For the case when you arm during erase

BLACKBOX_STATE_STOPPED -> BLACKBOX_STATE_START_ERASE -> BLACKBOX_STATE_ERASING -> *ARM* -> *Erase Finishes* -> BLACKBOX_STATE_PREPARE_LOG_FILE -> send all the headers -> BLACKBOX_STATE_RUNNING -> *Check for Blackbox switch* and either stay running or go BLACKBOX_STATE_PAUSED



Arming while erasing would allow you to only miss the first 0-20s of flight and get the rest recorded.  And usually that is what I would rather have since it is much more likely to actually catch an area where I would have needed the blackbox (rather then the log filling up and not getting any info)

brycedjohnson commented:
Bugfix and update: currently this pull won't start the blackbox log if is erasing during the arm.  The erase does complete successfully while arm though.

blckmn commented:
@ledvinap true - you get so used to sdcards... you forget :)

ledvinap commented:
@blckmn : If understand @thenickdude SD card implementation correctly, it makes sense to preallocate file with erased content on SDCard too ... But maybe it only avoids allocating new block for FAT (and possibility of double-sector erase)..

fftunes commented:
Nice. Confirmation beep would be perfect, i even suggested that a while ago.

Joshua Bardwell will like this too. :)

mikeller commented on src/main/blackbox/blackbox.c:
```diff
@@ -232,23 +232,21 @@ static const blackboxSimpleFieldDefinition_t blackboxSlowFields[] = {
+    BLACKBOX_STATE_RUNNING, //10
+    BLACKBOX_STATE_SHUTTING_DOWN, //11
+    BLACKBOX_STATE_START_ERASE, //12
+	BLACKBOX_STATE_ERASING, //13
```
Use spaces instead of tabs.

brycedjohnson replied:
Oops, I thought I turned that to spaces always in the editor...

mikeller commented on src/main/io/beeper.c:
```diff
@@ -171,9 +171,9 @@ typedef struct beeperTableEntry_s {
-
-    { BEEPER_ENTRY(BEEPER_ALL,                   18, NULL,                 "ALL") },
-    { BEEPER_ENTRY(BEEPER_PREFERENCE,            19, NULL,                 "PREFERRED") },
+    { BEEPER_ENTRY(BEEPER_BLACKBOX_ERASE,        18, beep_2shortBeeps,     "BLACKBOX_ERASE") },
```
Add the new entry at the end, renumbering will break things.

brycedjohnson replied:
Will do thanks!

brycedjohnson replied:
On that enum it says:     // BEEPER_ALL and BEEPER_PREFERENCE must remain at the bottom of this enum

mikeller replied:
Ah yes, that's the enum that uses id for priority ordering. We're not accessing beeper settings over MSP at the moment, so reordering should work.

mikeller commented:
Can you rebase and squash please?

brycedjohnson commented:
@mikeller So I shouldn't have done the merge there?  What is the best way to rebase?

brycedjohnson commented:
Added some changes so you only can erase once per reboot in case someone leaves the switch on.  Tested quite a few scenarios.   Also added in the beeps for erase start and finish.

Erase Switch -> Erase Complete -> Arm -> Start blackbox: works as expected

Erase Switch -> Erase Complete -> Blackbox Switch -> Arm: works as expected.  Starts recording at the Arm

Arm -> Erase:  Doesn't do anything because the erase has to be started in the BLACKBOX_STOPPED state

Erase -> Arm (Fly) -> Blackbox Switch -> Erase Complete:  Blackbox starts after the erase is done (while flying)

Erase -> Blackbox Switch -> Arm (Fly) -> Erase Complete: Same as above

Arm -> Disarm -> Erase -> Erase Complete -> Arm -> Blackbox Switch: Works as expected

mikeller commented:
No, the pull-up merge from master was all good. The rebase / squash is just cosmetics to make all of your changes appear as a single commit once they are merged into master.

To perform it, do `git rebase - i master`, then choose `pick` for one of your commits, `squash` for all others. Once that is done, use `git push --force-with-lease` (the `--force-with-lease` is crucial) to rewrite the history on your branch.

brycedjohnson commented:
@mikeller Ok I think I got it all in there.  Thanks for the help!  Everything seems to be working well for me, if anyone has any issues let me know.

mikeller commented:
@brycedjohnson: Hmm, sorry that I am so late with this, but one more thing: Since we are in a constant struggle against running out of flash space on F1 and F3 targets already, ideally all of the blackbox erasing functionality should be made conditional on `USE_FLASHFS`. This way, this change won't increase the size of targets that don't have flash memory, and hence won't be able to provide this functionality anyway.

Also, I will test your changes tonight.

brycedjohnson commented:
@mikeller I have some of that in there.  Such as this: https://github.com/betaflight/betaflight/pull/2267/commits/22d11e35b295087c696fa836141329f234d51e77#diff-8841562b3703006793b94c6ed1b03f8fR619

I didn't reduce it more because I thought some of the erase could be used for sdcards as well at some point?

As an example, your thinking blackboxEraseAll should be define on USE_FLASHFS as well as the calls to that?

mikeller commented:
Yes, that's my thinking. There are a number of factors that make me think erasing SDCard logs in the field is not a desirable option:
- with the way logs are written to SDCard, simply erasing a log file will not actually free up space to be written as log again. A lot of additional file system support would have to be built into the firmware for this;
- today's size / price of available SDCards allows you to add hours of logging capacity for a couple of USD, so running out of space is not a pressing issue;
- even if running out of space in the field is an issue, the easy workaround if using SDCard is 'replace the card with an empty one'.

mikeller commented:
Works great for me, I think this is a valuable capability. Merging for now, we can always make it conditional on `USE_FLASHFS` in a later PR.

mikeller commented:
After testing your patch in the field today (it rocks!), one question: Why did you add the limitation to one erase per boot? I tried to capture one particular strange behaviour pattern, and multiple record / erase cycles during one battery runtime would have been helpful.

brycedjohnson commented:
Great, I'm glad it was some use.  I was thinking it might be a little less error prone to only allow once.   If you leave the switch on it will keep on erasing back to back.   Maybe only allowing to erase once between arming might be better?

thenickdude commented:
I suggest that it enter an "ERASED" state after erasing is complete, and in order to return to the stopped state, the erase switch must be released.

mikeller commented:
I like @thenickdude's suggestion.
